lib/spack/spack/parser.py

```diff
@@ -390,27 +396,24 @@ def parse(self, initial_spec: spack.spec.Spec) -> spack.spec.Spec:
                 name = name.strip("'\" ")
                 value = value.strip("'\" ")
                 initial_spec._add_flag(name, value, propagate=True)
-            elif not self.has_hash and self.ctx.accept(TokenType.DAG_HASH):
-                dag_hash = self.ctx.current_token.value[1:]
-                matches = []
-                active_env = spack.environment.active_environment()
-                if active_env:
-                    matches = active_env.get_by_hash(dag_hash)
-                if not matches:
-                    matches = spack.store.db.get_by_hash(dag_hash)
-                if not matches:
-                    raise spack.spec.NoSuchHashError(dag_hash)
-
-                if len(matches) != 1:
-                    raise spack.spec.AmbiguousHashError(
-                        f"Multiple packages specify hash beginning '{dag_hash}'.", *matches
-                    )
-                spec_by_hash = matches[0]
-                if not spec_by_hash.satisfies(initial_spec):
-                    raise spack.spec.InvalidHashError(initial_spec, spec_by_hash.dag_hash())
-                initial_spec._dup(spec_by_hash)
-
-                self.has_hash = True
+            elif self.ctx.accept(TokenType.DAG_HASH):
+                # if initial_spec.abstract_hash:
+                #     msg = (
+                #         f"Parsed multiple hashes /{initial_spec.abstract_hash} and "
+                #         "/{self.current_token}. If you were attempting to specify a file path,"
+                #         " check that it was formatted properly"
+                #     )
+                #     raise spack.spec.RedundantSpecError(msg, self.ctx.current_token.value)
+                initial_spec.abstract_hash = self.ctx.current_token.value[1:]
+                if self.ctx.next_token and self.ctx.next_token.kind not in [
+                    TokenType.UNQUALIFIED_PACKAGE_NAME,
+                    TokenType.FULLY_QUALIFIED_PACKAGE_NAME,
+                    TokenType.DAG_HASH,
+                    TokenType.FILENAME,
+                    TokenType.DEPENDENCY,
+                ]:
```

Comment:
I would move this into a `TokenContxt.expect(*tokens: TokenType) -> bool` method, and call it, e.g.:

```suggestion
                if not self.ctx.expect(
                    TokenType.UNQUALIFIED_PACKAGE_NAME,
                    TokenType.FULLY_QUALIFIED_PACKAGE_NAME,
                    TokenType.DAG_HASH,
                    TokenType.FILENAME,
                    TokenType.DEPENDENCY,
                ):
```